File: C#.Net/JDI/Web/JDI Web/Selenium/Elements/Composite/WebPage.cs
using System;
using System.Linq;
using System.Text.RegularExpressions;
using Epam.JDI.Core.Interfaces.Complex;
using JDI_Commons;
using JDI_Web.Selenium.Base;
using JDI_Web.Settings;
using OpenQA.Selenium;
using RestSharp.Extensions;
using static System.String;
using static Epam.JDI.Core.ExceptionUtils;
using static Epam.JDI.Core.Settings.JDISettings;

namespace JDI_Web.Selenium.Elements.Composite
{
    public class WebPage : WebBaseElement, IPage
    {
        public static bool CheckAfterOpen = false;
        private string _url;

        public string Url
        {
            get => _url == null || _url.Contains("://") || !WebSettings.HasDomain
                ? _url
                : GetUrlFromUri(_url);
            set => _url = value;
        }

        public string Title;
        protected CheckPageTypes CheckUrlType = CheckPageTypes.None;
        protected CheckPageTypes CheckTitleType = CheckPageTypes.None;
        protected string UrlTemplate;
        public static WebPage CurrentPage;

        public WebPage()
        {
        }

        public WebPage(string url = null, string title = null)
        {
            Url = url;
            Title = title;
        }

        public static string GetUrlFromUri(string uri)
        {
            return WebSettings.Domain.Replace("/*$", "") + "/" + new Regex("^//*").Replace(uri, "");
        }

        public static string GetMatchFromDomain(string uri)
        {
            return WebSettings.Domain.Replace("/*$", "").Replace(".", "\\.") + "/" + uri.Replace("^/*", "");
        }

        public static void OpenUrl(string url)
        {
            new WebPage(url).Open();
        }

        public static string GetUrl()
        {
            return WebSettings.WebDriver.Url;
        }

        public static string GetTitle()
        {
            return WebSettings.WebDriver.Title;
        }

        public void UpdatePageData(string url, string title, CheckPageTypes checkUrlType, CheckPageTypes checkTitleType,
            string urlTemplate)
        {
            if (_url == null)
                Url = url;
            if (Title == null)
                Title = title;
            CheckUrlType = checkUrlType;
            CheckTitleType = checkTitleType;
            UrlTemplate = urlTemplate;
        }

        public StringCheckType CheckUrl()
        {
            return new StringCheckType(() => WebDriver.Url, Url, UrlTemplate, "url", () => Timer);
        }

        public StringCheckType CheckTitle()
        {
            return new StringCheckType(() => WebDriver.Title, Title, Title, "title", () => Timer);
        }

        public void CheckOpened()
        {
            if (IsNullOrEmpty(UrlTemplate) && new[] {CheckPageTypes.None, CheckPageTypes.Equal}.Contains(CheckUrlType))
                CheckUrl().Equal();
            else
                switch (CheckUrlType)
                {
                    case CheckPageTypes.None:
                        Asserter.IsTrue(GetUrl().Contains(UrlTemplate)
                                        || GetUrl().Matches(UrlTemplate));
                        break;
                    case CheckPageTypes.Equal:
                        CheckUrl().Equal();
                        break;
                    case CheckPageTypes.Match:
                        CheckUrl().Match();
                        break;
                    case CheckPageTypes.Contains:
                        CheckUrl().Contains();
                        break;
                }
            switch (CheckTitleType)
            {
                case CheckPageTypes.None:
                    CheckTitle().Equal();
                    break;
                case CheckPageTypes.Equal:
                    CheckTitle().Equal();
                    break;
                case CheckPageTypes.Match:
                    CheckTitle().Match();
                    break;
                case CheckPageTypes.Contains:
                    CheckTitle().Contains();
                    break;
            }
        }

        public void Open()
        {
            Invoker.DoJAction($"Open page {Name} by url {Url}",
                el => WebDriver.Navigate().GoToUrl(Url));
            if (CheckAfterOpen)
                CheckOpened();
            CurrentPage = this;
        }

        private bool IsOnPage()
        {
            var url = WebDriver.Url;
            if (IsNullOrEmpty(UrlTemplate)
                && new[] {CheckPageTypes.None, CheckPageTypes.Equal}.Contains(CheckUrlType))
                return url.Equals(Url);
            switch (CheckUrlType)
            {
                case CheckPageTypes.None:
                    return url.Contains(UrlTemplate) || url.Matches(UrlTemplate);
                case CheckPageTypes.Equal:
                    return url.Equals(Url);
                case CheckPageTypes.Match:
                    return url.Matches(UrlTemplate);
                case CheckPageTypes.Contains:
                    return url.Contains(IsNullOrEmpty(UrlTemplate) ? Url : UrlTemplate);
            }
            return false;
        }

        public void IsOpened()
        {
            ActionWithException(() =>
            {
                if (!IsOnPage())
                    Open();
                Logger.Info($"Page {Name} is opened");
            }, ex => $"Can't open page {Name}. Reason: {ex}");
        }

        /**
         * Refresh current page
         */

        public void Refresh()
        {
            Invoker.DoJAction($"Refresh page {Name}",
                el => WebDriver.Navigate().Refresh());
        }

        /**
         * Go back to previous page
         */

        public void Back()
        {
            Invoker.DoJAction("Go back to previous page",
                el => WebDriver.Navigate().Back());
        }


        /**
         * Go forward to next page
         */

        public void Forward()
        {
            Invoker.DoJAction("Go forward to next page",
                el => WebDriver.Navigate().Forward());
        }

        /**
         * @param cookie Specify cookie
         *               Add cookie in browser
         */

        public void AddCookie(Cookie cookie)
        {
            Invoker.DoJAction("Go forward to next page",
                el => WebDriver.Manage().Cookies.AddCookie(cookie));
        }

        /**
         * Clear browsers cache
         */

        public void ClearCache()
        {
            Invoker.DoJAction("Go forward to next page",
                el => WebDriver.Manage().Cookies.DeleteAllCookies());
        }

        public class StringCheckType
        {
            private readonly Func<string> _actual;
            private readonly string _equals;
            private readonly string _template;
            private readonly string _what;
            private readonly Func<Timer> _timer;

            public StringCheckType(Func<string> actual, string equals, string template, string what, Func<Timer> timer)
            {
                _actual = actual;
                _equals = equals;
                _template = template;
                _what = what;
                _timer = timer;
            }

            /**
             * Check that current page url/title equals to expected url/title
             */

            public void Equal()
            {
                if (IsNullOrEmpty(_equals)) return;
                Logger.Info($"Page {_what} equals to '{_equals}'");
                Asserter.IsTrue(_timer().Wait(() => _actual().Equals(_equals)));
            }

            /**
             * Check that current page url/title matches to expected url/title-matcher
             */

            public void Match()
            {
                if (IsNullOrEmpty(_template)) return;
                Logger.Info($"Page {_what} matches to '{_template}'");
                Asserter.IsTrue(_timer().Wait(() => _actual().Matches(_template)));
            }

            /**
             * Check that current page url/title contains expected url/title-matcher
             */

            public void Contains()
            {
                var url = IsNullOrEmpty(_template)
                    ? _equals
                    : _template;
                Logger.Info($"Page {_what} contains to '{url}'");
                Asserter.IsTrue(_timer().Wait(() => _actual().Contains(url)));
            }
        }
    }
}

File: Java/JDI/jdi-uitest-win/src/main/java/com/epam/jdi/uitests/win/winium/elements/composite/WebPage.java
package com.epam.jdi.uitests.win.winium.elements.composite;
/*
 * Copyright 2004-2016 EPAM Systems
 *
 * This file is part of JDI project.
 *
 * JDI is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Lesser General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * JDI is distributed in the hope that it will be useful, but WITHOUT ANY WARRANTY; 
 * without even the implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
 * See the GNU Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with JDI. If not, see <http://www.gnu.org/licenses/>.
 */


import com.epam.commons.Timer;
import com.epam.jdi.uitests.core.annotations.JDIAction;
import com.epam.jdi.uitests.core.interfaces.complex.IPage;
import com.epam.jdi.uitests.win.winium.elements.BaseElement;
import com.epam.jdi.uitests.win.settings.WinSettings;
import org.openqa.selenium.By;
import org.openqa.selenium.Cookie;

import java.util.function.Supplier;

import static com.epam.jdi.uitests.core.settings.JDISettings.*;
import static com.epam.jdi.uitests.win.settings.WinSettings.domain;
import static java.lang.String.format;

/**
 * Created by Roman_Iovlev on 7/17/2015.
 */
public class WebPage extends BaseElement implements IPage {
    public static boolean checkAfterOpen = false;
    public String url;
    public String title;
    protected CheckPageTypes checkUrlType = CheckPageTypes.EQUAL;
    protected CheckPageTypes checkTitleType = CheckPageTypes.EQUAL;
    protected String urlTemplate;
    public static WebPage currentPage;

    public WebPage() {
    }

    public WebPage(String url) {
        this.url = url;
    }

    public WebPage(String url, String title) {
        this.url = url;
        this.title = title;
    }

    public static String getUrlFromUri(String uri) {
        return domain.replaceAll("/*$", "") + "/" + uri.replaceAll("^/*", "");
    }

    public static String getMatchFromDomain(String uri) {
        return domain.replaceAll("/*$", "").replace(".", "\\.") + "/" + uri.replaceAll("^/*", "");
    }

    public static void openUrl(String url) {
        new WebPage(url).open();
    }

    public static String getUrl() {
        return WinSettings.getDriver().getCurrentUrl();
    }

    public static String getTitle() {
        return WinSettings.getDriver().getTitle();
    }

    public void updatePageData(String url, String title, CheckPageTypes checkUrlType, CheckPageTypes checkTitleType, String urlTemplate) {
        if (this.url == null)
            this.url = url;
        if (this.title == null)
            this.title = title;
        this.checkUrlType = checkUrlType;
        this.checkTitleType = checkTitleType;
        this.urlTemplate = urlTemplate;
    }

    public StringCheckType url() {
        return new StringCheckType(getDriver()::getCurrentUrl, url, urlTemplate, "url", timer());
    }

    public StringCheckType title() {
        return new StringCheckType(getDriver()::getTitle, title, title, "title", timer());
    }

    public void checkOpened() {
        switch (checkUrlType) {
            case EQUAL:
                url().check();
                break;
            case MATCH:
                url().match();
                break;
            case CONTAIN:
                url().contains();
                break;
        }
        switch (checkTitleType) {
            case EQUAL:
                title().check();
                break;
            case MATCH:
                title().match();
                break;
            case CONTAIN:
                title().contains();
                break;
        }
    }

    public <T extends IPage> T open() {
        invoker.doJAction(format("Open page '%s' by url %s", getName(), url),
                () -> getDriver().findElements(new By.ByClassName("Notepad")));
        if (checkAfterOpen)
            checkOpened();
        currentPage = this;
        return (T) this;
    }

    public void isOpened() {
        try {
            logger.info(format("Page '%s' should be opened", getName()));
            if (getDriver().getCurrentUrl().equals(url)) return;
            open();
        } catch (Exception ex) {
            throw exception(format("Can't open page '%s'. Reason: %s", getName(), ex.getMessage()));
        }
    }

    /**
     * Refresh current page
     */
    @JDIAction
    public void refresh() {
        invoker.doJAction(format("Refresh page '%s", getName()),
                () -> getDriver().navigate().refresh());
    }

    /**
     * Go back to previous page
     */
    @JDIAction
    public void back() {
        invoker.doJAction("Go back to previous page",
                () -> getDriver().navigate().back());
    }


    /**
     * Go forward to next page
     */
    @JDIAction
    public void forward() {
        invoker.doJAction("Go forward to next page",
                () -> getDriver().navigate().forward());
    }

    /**
     * @param cookie Specify cookie
     *               Add cookie in browser
     */
    @JDIAction
    public void addCookie(Cookie cookie) {
        invoker.doJAction("Go forward to next page",
                () -> getDriver().manage().addCookie(cookie));
    }

    /**
     * Clear browsers cache
     */
    @JDIAction
    public void clearCache() {
        invoker.doJAction("Go forward to next page",
                () -> getDriver().manage().deleteAllCookies());
    }

    public class StringCheckType {
        private Supplier<String> actual;
        private String equals;
        private String template;
        private String what;
        private Timer timer;

        public StringCheckType(Supplier<String> actual, String equals, String template, String what, Timer timer) {
            this.actual = actual;
            this.equals = equals;
            this.template = template;
            this.what = what;
            this.timer = timer;
        }

        /**
         * Check that current page url/title equals to expected url/title
         */
        @JDIAction
        public void check() {
            asserter.check(format("page %s equals to '%s'", what, equals))
                    .isTrue(timer.wait(() -> actual.get().equals(equals)));
        }

        /**
         * Check that current page url/title matches to expected url/title-matcher
         */
        @JDIAction
        public void match() {
            asserter.check(format("page %s matches to '%s'", what, template))
                    .isTrue(timer.wait(() -> actual.get().matches(template)));
        }

        /**
         * Check that current page url/title contains expected url/title-matcher
         */
        @JDIAction
        public void contains() {
            asserter.check(format("page %s contains '%s'", what, equals))
                    .isTrue(timer.wait(() -> actual.get().contains(template)));
        }
    }
}